pkg/controller/serviceaccount/tokens_controller.go

```diff
@@ -103,206 +113,269 @@ func NewTokensController(cl clientset.Interface, options TokensControllerOptions
-func (e *TokensController) Run() {
-	if e.stopChan == nil {
-		e.stopChan = make(chan struct{})
-		go e.serviceAccountController.Run(e.stopChan)
-		go e.secretController.Run(e.stopChan)
+// Runs controller blocks until stopCh is closed
+func (e *TokensController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+
+	// Start controllers (to fill stores, call informers, fill work queues)
+	go e.serviceAccountController.Run(stopCh)
+	go e.secretController.Run(stopCh)
+
+	// Wait for stores to fill
+	for !e.serviceAccountController.HasSynced() || !e.secretController.HasSynced() {
+		time.Sleep(100 * time.Millisecond)
 	}
-}
 
-// Stop gracefully shuts down this controller
-func (e *TokensController) Stop() {
-	if e.stopChan != nil {
-		close(e.stopChan)
-		e.stopChan = nil
+	// Spawn workers to process work queues
+	e.activeWorkers = workers
+	for i := 0; i < workers; i++ {
+		go wait.Until(e.syncServiceAccount, 0, stopCh)
+		go wait.Until(e.syncSecret, 0, stopCh)
 	}
+
+	// Block until stop channel is closed
+	<-stopCh
+
+	// Shut down queues
+	e.syncServiceAccountQueue.ShutDown()
+	e.syncSecretQueue.ShutDown()
 }
 
-// serviceAccountAdded reacts to a ServiceAccount creation by creating a corresponding ServiceAccountToken Secret
-func (e *TokensController) serviceAccountAdded(obj interface{}) {
-	serviceAccount := obj.(*api.ServiceAccount)
-	err := e.createSecretIfNeeded(serviceAccount)
-	if err != nil {
-		glog.Error(err)
+func (e *TokensController) queueServiceAccountSync(obj interface{}) {
+	if serviceAccount, ok := obj.(*api.ServiceAccount); ok {
+		// glog.Infof("queueServiceAccountSync: %s/%s", serviceAccount.Namespace, serviceAccount.Name)
+		e.syncServiceAccountQueue.Add(makeServiceAccountKey(serviceAccount))
 	}
 }
 
-// serviceAccountUpdated reacts to a ServiceAccount update (or re-list) by ensuring a corresponding ServiceAccountToken Secret exists
-func (e *TokensController) serviceAccountUpdated(oldObj interface{}, newObj interface{}) {
-	newServiceAccount := newObj.(*api.ServiceAccount)
-	err := e.createSecretIfNeeded(newServiceAccount)
-	if err != nil {
-		glog.Error(err)
+func (e *TokensController) queueServiceAccountUpdateSync(oldObj interface{}, newObj interface{}) {
+	if serviceAccount, ok := newObj.(*api.ServiceAccount); ok {
+		// glog.Infof("queueServiceAccountUpdateSync: %s/%s", serviceAccount.Namespace, serviceAccount.Name)
+		e.syncServiceAccountQueue.Add(makeServiceAccountKey(serviceAccount))
 	}
 }
 
-// serviceAccountDeleted reacts to a ServiceAccount deletion by deleting all corresponding ServiceAccountToken Secrets
-func (e *TokensController) serviceAccountDeleted(obj interface{}) {
-	serviceAccount, ok := obj.(*api.ServiceAccount)
-	if !ok {
-		// Unknown type. If we missed a ServiceAccount deletion, the
-		// corresponding secrets will be cleaned up during the Secret re-list
-		return
+func (e *TokensController) requeue(queue *workqueue.Type, key interface{}) {
```

Comment:
> #23444

That in combination with a `ratelimit.NewBucketWithRate()` and then `bucket.Take()` is what I planned to do downstream.  I think that should give the duration to add to the requeue.  MaxFailures should be tracked both on your object (which may fail) and a list of those failures in your controller (which should not fail).